File: java/Samples/BiometricStandards/CS/ANTemplateSampleCS/FieldForm.cs
using System;
using System.Text;
using System.Windows.Forms;
using Neurotec.Biometrics.Standards;

namespace Neurotec.Samples
{
	public partial class FieldForm : Form
	{
		#region Public static methods

		public static void GetFieldValue(ANField field, StringBuilder value)
		{
			value.Length = 0;
			bool manySubFields = field.SubFields.Count > 1;
			int sfi = 0;
			foreach (ANSubField subField in field.SubFields)
			{
				if (sfi != 0) value.Append(',');
				if (manySubFields) value.Append('{');
				int ii = 0;
				foreach (string item in subField.Items)
				{
					if (ii != 0) value.Append('|');
					value.Append(item);
					ii++;
				}
				if (manySubFields) value.Append('}');
				sfi++;
			}
		}

		#endregion

		#region Private fields

		private ANField _field;
		private bool _isReadOnly;
		private bool _isModified;

		#endregion

		#region Public constructor

		public FieldForm()
		{
			InitializeComponent();

			OnFieldChanged();
		}

		#endregion

		#region Private methods

		private void UpdateField()
		{
			if (_field != null)
			{
				StringBuilder value = new StringBuilder();
				GetFieldValue(_field, value);
				fieldValueLabel.Text = value.ToString();
			}
			else
			{
				fieldValueLabel.Text = string.Empty;
			}
			editFieldButton.Enabled = _field != null && _field.SubFields.Count == 1 && _field.SubFields[0].Items.Count == 1;
			_isModified = true;
		}

		private void OnFieldChanged()
		{
			subFieldListBox.BeginUpdate();
			subFieldListBox.Items.Clear();
			if (_field != null)
			{
				StringBuilder value = new StringBuilder();
				foreach (ANSubField subField in _field.SubFields)
				{
					GetSubFieldValue(subField, value);
					subFieldListBox.Items.Add(value.ToString());
				}
				subFieldListBox.SelectedIndex = 0;
			}
			subFieldListBox.EndUpdate();
			UpdateField();
			_isModified = false;
		}

		private void EditField()
		{
			ItemForm form = new ItemForm();
			form.Value = _field.Value;
			form.Text = "Edit Field";
			form.IsReadOnly = _isReadOnly;
			if (form.ShowDialog() == DialogResult.OK)
			{
				_field.Value = form.Value;
				UpdateSelectedSubField(); // couse we need to update subfield also
			}
		}

		private void OnIsReadOnlyChanged()
		{
			UpdateSubFieldControls();
			UpdateItemControls();
		}

		private ANSubField GetSelectedSubField()
		{
			return subFieldListBox.SelectedIndices.Count == 1 ? _field.SubFields[subFieldListBox.SelectedIndex] : null;
		}

		private void GetSubFieldValue(ANSubField subField, StringBuilder value)
		{
			value.Length = 0;
			int ii = 0;
			foreach (string item in subField.Items)
			{
				if (ii != 0) value.Append('|');
				value.Append(item);
				ii++;
			}
		}

		private void OnSelectedSubFieldChanged()
		{
			ANSubField selectedSubField = GetSelectedSubField();
			itemListBox.BeginUpdate();
			itemListBox.Items.Clear();
			if (selectedSubField != null)
			{
				foreach (string item in selectedSubField.Items)
				{
					itemListBox.Items.Add(item);
				}
				itemListBox.SelectedIndex = 0;
			}
			itemListBox.EndUpdate();
			UpdateSubFieldControls();
		}

		private void UpdateSubFieldControls()
		{
			ANSubField selectedSubField = GetSelectedSubField();
			addSubFieldButton.Enabled = _field != null && !_isReadOnly;
			insertSubFieldButton.Enabled = !_isReadOnly && selectedSubField != null;
			int sc = subFieldListBox.SelectedIndices.Count;
			removeSubFieldButton.Enabled = !_isReadOnly && sc != 0 && sc != _field.SubFields.Count;
			editSubFieldButton.Enabled = selectedSubField != null && selectedSubField.Items.Count == 1;
		}

		private void UpdateSubField(int index)
		{
			StringBuilder subFieldValue = new StringBuilder();
			GetSubFieldValue(_field.SubFields[index], subFieldValue);
			subFieldListBox.Items[index] = subFieldValue.ToString();
			UpdateField();
		}

		private void UpdateSelectedSubField()
		{
			UpdateSubField(subFieldListBox.SelectedIndex);
		}

		private void EditSubField()
		{
			ANSubField selectedSubField = GetSelectedSubField();
			ItemForm form = new ItemForm();
			form.Value = selectedSubField.Value;
			form.Text = "Edit Subfield";
			form.IsReadOnly = _isReadOnly;
			if (form.ShowDialog() == DialogResult.OK)
			{
				selectedSubField.Value = form.Value;
				UpdateSelectedSubField();
			}
		}

		private int GetSelectedItemIndex()
		{
			return itemListBox.SelectedIndices.Count == 1 ? itemListBox.SelectedIndex : -1;
		}

		private void OnSelectedItemChanged()
		{
			UpdateItemControls();
		}

		private void UpdateItemControls()
		{
			ANSubField selectedSubField = GetSelectedSubField();
			int selectedItemIndex = GetSelectedItemIndex();
			addItemButton.Enabled = selectedSubField != null && !_isReadOnly;
			insertItemButton.Enabled = !_isReadOnly && selectedItemIndex != -1;
			int sc = itemListBox.SelectedIndices.Count;
			removeItemButton.Enabled = selectedSubField != null && !_isReadOnly && sc != 0 && sc != selectedSubField.Items.Count;
			editItemButton.Enabled = selectedItemIndex != -1;
		}

		private void EditItem()
		{
			ANSubField selectedSubField = GetSelectedSubField();
			int selectedItemIndex = GetSelectedItemIndex();
			ItemForm form = new ItemForm();
			form.Value = selectedSubField.Items[selectedItemIndex];
			form.IsReadOnly = _isReadOnly;
			if (form.ShowDialog() == DialogResult.OK)
			{
				selectedSubField.Items[selectedItemIndex] = form.Value;
				UpdateSelectedSubField();
				itemListBox.SelectedIndex = -1;
				itemListBox.SelectedIndex = selectedItemIndex;
			}
		}

		#endregion

		#region Public properties

		public ANField Field
		{
			get
			{
				return _field;
			}
			set
			{
				if (_field != value)
				{
					_field = value;
					OnFieldChanged();
				}
			}
		}

		public bool IsModified
		{
			get
			{
				return _isModified;
			}
		}

		public bool IsReadOnly
		{
			get
			{
				return _isReadOnly;
			}
			set
			{
				if (_isReadOnly != value)
				{
					_isReadOnly = value;
					OnIsReadOnlyChanged();
				}
			}
		}

		#endregion

		#region Private form events

		private void FieldValueLabelDoubleClick(object sender, EventArgs e)
		{
			if (_field != null && _field.SubFields.Count == 1 && _field.SubFields[0].Items.Count == 1)
			{
				EditField();
			}
		}

		private void EditFieldButtonClick(object sender, EventArgs e)
		{
			EditField();
		}

		private void SubFieldListBoxSelectedIndexChanged(object sender, EventArgs e)
		{
			OnSelectedSubFieldChanged();
		}

		private void SubFieldListBoxDoubleClick(object sender, EventArgs e)
		{
			ANSubField selectedSubField = GetSelectedSubField();
			if (selectedSubField != null && selectedSubField.Items.Count == 1)
			{
				EditSubField();
			}
		}

		private void AddSubFieldButtonClick(object sender, EventArgs e)
		{
			ItemForm form = new ItemForm();
			form.Text = "Add Subfield";
			if (form.ShowDialog() == DialogResult.OK)
			{
				_field.SubFields.Add(form.Value);
				subFieldListBox.Items.Add(string.Empty);
				int index = _field.SubFields.Count - 1;
				UpdateSubField(index);
				subFieldListBox.SelectedIndex = -1;
				subFieldListBox.SelectedIndex = index;
			}
		}

		private void InsertSubFieldButtonClick(object sender, EventArgs e)
		{
			int index = subFieldListBox.SelectedIndex;
			ItemForm form = new ItemForm();
			form.Text = "Insert Subfield";
			if (form.ShowDialog() == DialogResult.OK)
			{
				_field.SubFields.Insert(index, form.Value);
				subFieldListBox.Items.Insert(index, string.Empty);
				UpdateSubField(index);
				subFieldListBox.SelectedIndex = -1;
				subFieldListBox.SelectedIndex = index;
			}
		}

		private void RemoveSubFieldButtonClick(object sender, EventArgs e)
		{
			subFieldListBox.BeginUpdate();
			int selCount = subFieldListBox.SelectedIndices.Count;
			int[] selIndices = new int[selCount];
			subFieldListBox.SelectedIndices.CopyTo(selIndices, 0);
			Array.Sort<int>(selIndices);
			for (int i = selCount - 1; i >= 0; i--)
			{
				int index = selIndices[i];
				_field.SubFields.RemoveAt(index);
				subFieldListBox.Items.RemoveAt(index);
			}
			UpdateField();
			subFieldListBox.SelectedIndex = -1;
			subFieldListBox.SelectedIndex = selIndices[0] == _field.SubFields.Count ? _field.SubFields.Count - 1 : selIndices[0];
			subFieldListBox.EndUpdate();
		}

		private void EditSubFieldButtonClick(object sender, EventArgs e)
		{
			EditSubField();
		}

		private void ItemListBoxSelectedIndexChanged(object sender, EventArgs e)
		{
			OnSelectedItemChanged();
		}

		private void ItemListBoxDoubleClick(object sender, EventArgs e)
		{
			if (GetSelectedItemIndex() != -1)
			{
				EditItem();
			}
		}

		private void AddItemButtonClick(object sender, EventArgs e)
		{
			ANSubField selectedSubField = GetSelectedSubField();
			ItemForm form = new ItemForm();
			form.Text = "Add Item";
			if (form.ShowDialog() == DialogResult.OK)
			{
				int index = selectedSubField.Items.Add(form.Value);
				UpdateSelectedSubField();
				itemListBox.SelectedIndex = -1;
				itemListBox.SelectedIndex = index;
			}
		}

		private void InsertItemButtonClick(object sender, EventArgs e)
		{
			ANSubField selectedSubField = GetSelectedSubField();
			ItemForm form = new ItemForm();
			form.Text = "Insert Item";
			if (form.ShowDialog() == DialogResult.OK)
			{
				int index = GetSelectedItemIndex();
				selectedSubField.Items.Insert(index, form.Value);
				UpdateSelectedSubField();
				itemListBox.SelectedIndex = -1;
				itemListBox.SelectedIndex = index;
			}
		}

		private void RemoveItemButtonClick(object sender, EventArgs e)
		{
			itemListBox.BeginUpdate();
			ANSubField selectedSubField = GetSelectedSubField();
			int selCount = itemListBox.SelectedIndices.Count;
			int[] selIndices = new int[selCount];
			itemListBox.SelectedIndices.CopyTo(selIndices, 0);
			Array.Sort<int>(selIndices);
			for (int i = selCount - 1; i >= 0; i--)
			{
				int index = selIndices[i];
				selectedSubField.Items.RemoveAt(index);
			}
			UpdateSelectedSubField();
			itemListBox.SelectedIndex = -1;
			itemListBox.SelectedIndex = selIndices[0] == selectedSubField.Items.Count ? selectedSubField.Items.Count - 1 : selIndices[0];
			itemListBox.EndUpdate();
		}

		private void EditItemButtonClick(object sender, EventArgs e)
		{
			EditItem();
		}

		private void FieldFormShown(object sender, EventArgs e)
		{
			if (_field != null && _field.SubFields.Count == 1 && _field.SubFields[0].Items.Count == 1 && !IsReadOnly)
			{
				EditField();
			}
		}

		#endregion
	}
}

File: java/Samples/BiometricStandards/VB/ANTemplateSampleVB/FieldForm.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Text
Imports System.Windows.Forms
Imports Neurotec.Biometrics.Standards

Partial Public Class FieldForm
	Inherits Form
	#Region "Public static methods"

	Public Shared Sub GetFieldValue(ByVal field As ANField, ByVal value As StringBuilder)
		value.Length = 0
		Dim manySubFields As Boolean = field.SubFields.Count > 1
		Dim sfi As Integer = 0
		For Each subField As ANSubField In field.SubFields
			If sfi <> 0 Then
				value.Append(","c)
			End If
			If manySubFields Then
				value.Append("{"c)
			End If
			Dim ii As Integer = 0
			For Each item As String In subField.Items
				If ii <> 0 Then
					value.Append("|"c)
				End If
				value.Append(item)
				ii += 1
			Next item
			If manySubFields Then
				value.Append("}"c)
			End If
			sfi += 1
		Next subField
	End Sub

	#End Region

	#Region "Private fields"

	Private _field As ANField
	Private _isReadOnly As Boolean
	Private _isModified As Boolean

	#End Region

	#Region "Public constructor"

	Public Sub New()
		InitializeComponent()

		OnFieldChanged()
	End Sub

	#End Region

	#Region "Private methods"

	Private Sub UpdateField()
		If _field IsNot Nothing Then
			Dim value As New StringBuilder()
			GetFieldValue(_field, value)
			fieldValueLabel.Text = value.ToString()
		Else
			fieldValueLabel.Text = String.Empty
		End If
		editFieldButton.Enabled = _field IsNot Nothing AndAlso _field.SubFields.Count = 1 AndAlso _field.SubFields(0).Items.Count = 1
		_isModified = True
	End Sub

	Private Sub OnFieldChanged()
		subFieldListBox.BeginUpdate()
		subFieldListBox.Items.Clear()
		If _field IsNot Nothing Then
			Dim value As New StringBuilder()
			For Each subField As ANSubField In _field.SubFields
				GetSubFieldValue(subField, value)
				subFieldListBox.Items.Add(value.ToString())
			Next subField
			subFieldListBox.SelectedIndex = 0
		End If
		subFieldListBox.EndUpdate()
		UpdateField()
		_isModified = False
	End Sub

	Private Sub EditField()
		Dim form As New ItemForm()
		form.Value = _field.Value
		form.Text = "Edit Field"
		form.IsReadOnly = _isReadOnly
		If form.ShowDialog() = Windows.Forms.DialogResult.OK Then
			_field.Value = form.Value
			UpdateSelectedSubField() ' couse we need to update subfield also
		End If
	End Sub

	Private Sub OnIsReadOnlyChanged()
		UpdateSubFieldControls()
		UpdateItemControls()
	End Sub

	Private Function GetSelectedSubField() As ANSubField
		If subFieldListBox.SelectedIndices.Count = 1 Then
			Return _field.SubFields(subFieldListBox.SelectedIndex)
		Else
			Return Nothing
		End If
	End Function

	Private Sub GetSubFieldValue(ByVal subField As ANSubField, ByVal value As StringBuilder)
		value.Length = 0
		Dim ii As Integer = 0
		For Each item As String In subField.Items
			If ii <> 0 Then
				value.Append("|"c)
			End If
			value.Append(item)
			ii += 1
		Next item
	End Sub

	Private Sub OnSelectedSubFieldChanged()
		Dim selectedSubField As ANSubField = GetSelectedSubField()
		itemListBox.BeginUpdate()
		itemListBox.Items.Clear()
		If selectedSubField IsNot Nothing Then
			For Each item As String In selectedSubField.Items
				itemListBox.Items.Add(item)
			Next item
			itemListBox.SelectedIndex = 0
		End If
		itemListBox.EndUpdate()
		UpdateSubFieldControls()
	End Sub

	Private Sub UpdateSubFieldControls()
		Dim selectedSubField As ANSubField = GetSelectedSubField()
		addSubFieldButton.Enabled = _field IsNot Nothing AndAlso Not _isReadOnly
		insertSubFieldButton.Enabled = (Not _isReadOnly) AndAlso selectedSubField IsNot Nothing
		Dim sc As Integer = subFieldListBox.SelectedIndices.Count
		removeSubFieldButton.Enabled = (Not _isReadOnly) AndAlso sc <> 0 AndAlso sc <> _field.SubFields.Count
		editSubFieldButton.Enabled = selectedSubField IsNot Nothing AndAlso selectedSubField.Items.Count = 1
	End Sub

	Private Sub UpdateSubField(ByVal index As Integer)
		Dim subFieldValue As New StringBuilder()
		GetSubFieldValue(_field.SubFields(index), subFieldValue)
		subFieldListBox.Items(index) = subFieldValue.ToString()
		UpdateField()
	End Sub

	Private Sub UpdateSelectedSubField()
		UpdateSubField(subFieldListBox.SelectedIndex)
	End Sub

	Private Sub EditSubField()
		Dim selectedSubField As ANSubField = GetSelectedSubField()
		Dim form As New ItemForm()
		form.Value = selectedSubField.Value
		form.Text = "Edit Subfield"
		form.IsReadOnly = _isReadOnly
		If form.ShowDialog() = Windows.Forms.DialogResult.OK Then
			selectedSubField.Value = form.Value
			UpdateSelectedSubField()
		End If
	End Sub

	Private Function GetSelectedItemIndex() As Integer

		Return CInt(IIf(itemListBox.SelectedIndices.Count = 1, itemListBox.SelectedIndex, -1))
	End Function

	Private Sub OnSelectedItemChanged()
		UpdateItemControls()
	End Sub

	Private Sub UpdateItemControls()
		Dim selectedSubField As ANSubField = GetSelectedSubField()
		Dim selectedItemIndex As Integer = GetSelectedItemIndex()
		addItemButton.Enabled = selectedSubField IsNot Nothing AndAlso Not _isReadOnly
		insertItemButton.Enabled = (Not _isReadOnly) AndAlso selectedItemIndex <> -1
		Dim sc As Integer = itemListBox.SelectedIndices.Count
		removeItemButton.Enabled = (Not _isReadOnly) AndAlso sc <> 0 AndAlso sc <> selectedSubField.Items.Count
		editItemButton.Enabled = selectedItemIndex <> -1
	End Sub

	Private Sub EditItem()
		Dim selectedSubField As ANSubField = GetSelectedSubField()
		Dim selectedItemIndex As Integer = GetSelectedItemIndex()
		Dim form As New ItemForm()
		form.Value = selectedSubField.Items(selectedItemIndex)
		form.IsReadOnly = _isReadOnly
		If form.ShowDialog() = Windows.Forms.DialogResult.OK Then
			selectedSubField.Items(selectedItemIndex) = form.Value
			UpdateSelectedSubField()
			itemListBox.SelectedIndex = -1
			itemListBox.SelectedIndex = selectedItemIndex
		End If
	End Sub

	#End Region

	#Region "Public properties"

	Public Property Field() As ANField
		Get
			Return _field
		End Get
		Set(ByVal value As ANField)
			If _field IsNot value Then
				_field = value
				OnFieldChanged()
			End If
		End Set
	End Property

	Public ReadOnly Property IsModified() As Boolean
		Get
			Return _isModified
		End Get
	End Property

	Public Property IsReadOnly() As Boolean
		Get
			Return _isReadOnly
		End Get
		Set(ByVal value As Boolean)
			If _isReadOnly <> value Then
				_isReadOnly = value
				OnIsReadOnlyChanged()
			End If
		End Set
	End Property

	#End Region

	#Region "Private form events"

	Private Sub FieldValueLabelDoubleClick(ByVal sender As Object, ByVal e As EventArgs) Handles fieldValueLabel.DoubleClick
		If _field IsNot Nothing AndAlso _field.SubFields.Count = 1 AndAlso _field.SubFields(0).Items.Count = 1 Then
			EditField()
		End If
	End Sub

	Private Sub EditFieldButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles editFieldButton.Click
		EditField()
	End Sub

	Private Sub SubFieldListBoxSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles subFieldListBox.SelectedIndexChanged
		OnSelectedSubFieldChanged()
	End Sub

	Private Sub SubFieldListBoxDoubleClick(ByVal sender As Object, ByVal e As EventArgs) Handles subFieldListBox.DoubleClick
		Dim selectedSubField As ANSubField = GetSelectedSubField()
		If selectedSubField IsNot Nothing AndAlso selectedSubField.Items.Count = 1 Then
			EditSubField()
		End If
	End Sub

	Private Sub AddSubFieldButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles addSubFieldButton.Click
		Dim form As New ItemForm()
		form.Text = "Add Subfield"
		If form.ShowDialog() = Windows.Forms.DialogResult.OK Then
			_field.SubFields.Add(form.Value)
			subFieldListBox.Items.Add(String.Empty)
			Dim index As Integer = _field.SubFields.Count - 1
			UpdateSubField(index)
			subFieldListBox.SelectedIndex = -1
			subFieldListBox.SelectedIndex = index
		End If
	End Sub

	Private Sub InsertSubFieldButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles insertSubFieldButton.Click
		Dim index As Integer = subFieldListBox.SelectedIndex
		Dim form As New ItemForm()
		form.Text = "Insert Subfield"
		If form.ShowDialog() = Windows.Forms.DialogResult.OK Then
			_field.SubFields.Insert(index, form.Value)
			subFieldListBox.Items.Insert(index, String.Empty)
			UpdateSubField(index)
			subFieldListBox.SelectedIndex = -1
			subFieldListBox.SelectedIndex = index
		End If
	End Sub

	Private Sub RemoveSubFieldButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles removeSubFieldButton.Click
		subFieldListBox.BeginUpdate()
		Dim selCount As Integer = subFieldListBox.SelectedIndices.Count
		Dim selIndices(selCount - 1) As Integer
		subFieldListBox.SelectedIndices.CopyTo(selIndices, 0)
		Array.Sort(Of Integer)(selIndices)
		For i As Integer = selCount - 1 To 0 Step -1
			Dim index As Integer = selIndices(i)
			_field.SubFields.RemoveAt(index)
			subFieldListBox.Items.RemoveAt(index)
		Next i
		UpdateField()
		subFieldListBox.SelectedIndex = -1
		If selIndices(0) = _field.SubFields.Count Then
			subFieldListBox.SelectedIndex = _field.SubFields.Count - 1
		Else
			subFieldListBox.SelectedIndex = selIndices(0)
		End If
		subFieldListBox.EndUpdate()
	End Sub

	Private Sub EditSubFieldButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles editSubFieldButton.Click
		EditSubField()
	End Sub

	Private Sub ItemListBoxSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles itemListBox.SelectedIndexChanged
		OnSelectedItemChanged()
	End Sub

	Private Sub ItemListBoxDoubleClick(ByVal sender As Object, ByVal e As EventArgs) Handles itemListBox.DoubleClick
		If GetSelectedItemIndex() <> -1 Then
			EditItem()
		End If
	End Sub

	Private Sub AddItemButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles addItemButton.Click
		Dim selectedSubField As ANSubField = GetSelectedSubField()
		Dim form As New ItemForm()
		form.Text = "Add Item"
		If form.ShowDialog() = Windows.Forms.DialogResult.OK Then
			Dim index As Integer = selectedSubField.Items.Add(form.Value)
			UpdateSelectedSubField()
			itemListBox.SelectedIndex = -1
			itemListBox.SelectedIndex = index
		End If
	End Sub

	Private Sub InsertItemButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles insertItemButton.Click
		Dim selectedSubField As ANSubField = GetSelectedSubField()
		Dim form As New ItemForm()
		form.Text = "Insert Item"
		If form.ShowDialog() = Windows.Forms.DialogResult.OK Then
			Dim index As Integer = GetSelectedItemIndex()
			selectedSubField.Items.Insert(index, form.Value)
			UpdateSelectedSubField()
			itemListBox.SelectedIndex = -1
			itemListBox.SelectedIndex = index
		End If
	End Sub

	Private Sub RemoveItemButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles removeItemButton.Click
		itemListBox.BeginUpdate()
		Dim selectedSubField As ANSubField = GetSelectedSubField()
		Dim selCount As Integer = itemListBox.SelectedIndices.Count
		Dim selIndices(selCount - 1) As Integer
		itemListBox.SelectedIndices.CopyTo(selIndices, 0)
		Array.Sort(Of Integer)(selIndices)
		For i As Integer = selCount - 1 To 0 Step -1
			Dim index As Integer = selIndices(i)
			selectedSubField.Items.RemoveAt(index)
		Next i
		UpdateSelectedSubField()
		itemListBox.SelectedIndex = -1
		If selIndices(0) = selectedSubField.Items.Count Then
			itemListBox.SelectedIndex = selectedSubField.Items.Count - 1
		Else
			itemListBox.SelectedIndex = selIndices(0)
		End If
		itemListBox.EndUpdate()
	End Sub

	Private Sub EditItemButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles editItemButton.Click
		EditItem()
	End Sub

	Private Sub FieldFormShown(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Shown
		If _field IsNot Nothing AndAlso _field.SubFields.Count = 1 AndAlso _field.SubFields(0).Items.Count = 1 AndAlso (Not IsReadOnly) Then
			EditField()
		End If
	End Sub

	#End Region
End Class
